module/zfs/vdev_raidz.c

```diff
@@ -1525,41 +1538,71 @@ vdev_raidz_io_start_write(zio_t *zio, raidz_row_t *rr, uint64_t ashift)
 {
 	vdev_t *vd = zio->io_vd;
 	raidz_map_t *rm = zio->io_vsd;
-	int c, i;
+	int c, skipped = 0, skip_first_cols = 0;
 
 	vdev_raidz_generate_parity_row(rm, rr);
 
-	for (int c = 0; c < rr->rr_cols; c++) {
+	if (rr->rr_scols < (rm->rm_skipstart + rm->rm_nskip)) {
+		skip_first_cols =
+		    (rm->rm_skipstart + rm->rm_nskip) % rr->rr_scols;
+	}
+
+	for (c = 0; c < rr->rr_scols; c++) {
+		abd_t *abd = NULL;
+		enum zio_flag flags = 0;
 		raidz_col_t *rc = &rr->rr_col[c];
-		if (rc->rc_size == 0)
-			continue;
+		vdev_t *cvd = vd->vdev_child[rc->rc_devidx];
 
 		/* Verify physical to logical translation */
 		vdev_raidz_io_verify(vd, rr, c);
 
-		zio_nowait(zio_vdev_child_io(zio, NULL,
-		    vd->vdev_child[rc->rc_devidx], rc->rc_offset,
-		    rc->rc_abd, rc->rc_size, zio->io_type, zio->io_priority,
-		    0, vdev_raidz_child_done, rc));
-	}
+		/*
+		 * Generate I/O for skip sectors to improve aggregation
+		 * contiguity. We will use gang ABD's to reduce contention
+		 * on the children VDEV queue locks (vq_lock) by issuing
+		 * a single I/O that contains the data and skip sectors.
+		 */
+		if (c < skip_first_cols || (c >= rm->rm_skipstart &&
+		    skipped < rm->rm_nskip)) {
+			if (rc->rc_size > 0) {
+				abd = abd_alloc_gang();
+				abd_gang_add(abd, rc->rc_abd, B_FALSE);
+			} else {
+				ASSERT3P(rc->rc_abd, ==, NULL);
+				zio_nowait(zio_vdev_child_io(zio, NULL, cvd,
+				    rc->rc_offset, NULL, 1ULL << ashift,
+				    zio->io_type, zio->io_priority,
+				    ZIO_FLAG_NODATA | ZIO_FLAG_OPTIONAL, NULL,
+				    NULL));
+				skipped++;
+				continue;
+			}
+			abd_gang_add(abd, abd_get_zeros(1ULL << ashift),
+			    B_TRUE);
 
-	/*
-	 * Generate optional I/Os for skip sectors to improve aggregation
-	 * contiguity.
-	 */
-	for (c = rm->rm_skipstart, i = 0; i < rm->rm_nskip; c++, i++) {
-		ASSERT(c <= rr->rr_scols);
-		if (c == rr->rr_scols)
-			c = 0;
+			/*
+			 * Store original ABD so the gang ABD can be freed in
+			 * vdev_raidz_child_done().
+			 */
+			ASSERT3P(rc->rc_orig_data, ==, NULL);
+			rc->rc_orig_data = rc->rc_abd;
```

Comment:
Is the reason it's safe to reuse `rc->rc_orig_data` because it's only used during reconstruction which will only happen during a read, thus it's always unused for writes.  If that's the case could to update the comment to include this rational.

Reply:
I updated the comment. I also wound up moving all of this (including the comment) into the `if` block as well to make it more readable.